[4.4] Fix handling of sub-ms transaction timeouts

robsdedude opened the pull request:
Transaction timeouts are specified in seconds as float. However, the server expects it in milliseconds as int. This would lead to

 1) rounding issues: previously, the driver would multiply by 1000 and
    then truncate to int. E.g., 256.4 seconds would be turned into 256399 ms
    because of float imprecision.
    Therefore, the built-in `round` is now used instead.
 2) values below 1 ms (e.g., 0.0001) would be rounded down to 0. However, 0 is
    a special value that instructs the server to not apply any timeout. This
    is likely to surprise the user which specified a non-zero timeout. In this
    special case, the driver now rounds up to 1 ms.

Back-port of: https://github.com/neo4j/neo4j-python-driver/pull/940

robsdedude committed:
Fix handling of sub-ms transaction timeouts

Transaction timeouts are specified in seconds as float.
However, the server expects it in milliseconds as int. This would lead to

 1) rounding issues: previously, the driver would multiply by 1000 and
    then truncate to int. E.g., 256.4 seconds would be turned into 256399 ms
    because of float imprecision.
    Therefore, the built-in `round` is now used instead.
 2) values below 1 ms (e.g., 0.0001) would be rounded down to 0. However, 0 is
    a special value that instructs the server to not apply any timeout. This
    is likely to surprise the user which specified a non-zero timeout. In this
    special case, the driver now rounds up to 1 ms.

Back-port of: https://github.com/neo4j/neo4j-python-driver/pull/940

AndyHeap-NeoTech commented on neo4j/io/_bolt3.py:
```diff
@@ -225,11 +226,8 @@ def run(self, query, parameters=None, mode=None, bookmarks=None,
-                extra["tx_timeout"] = int(1000 * timeout)
-            except TypeError:
-                raise TypeError("Timeout must be specified as a number of seconds")
+        if timeout or timeout == 0:
```
Not sure if I'm reading this correctly. Is timeout a float here and being used in an equality check?

robsdedude replied:
I applied a small code change, but it can still be a comparison of a float (depending on what type the user passes to the driver). However, to my knowledge, only the only float that's falsy is `float(0)`. So the first half half of the condition would short circuit for any other value.

AndyHeap-NeoTech reviewed (approved):
Just one question. Looks good though.

robsdedude committed:
Improve backwards-compatibility

robsdedude committed:
Improve API docs

robsdedude committed:
Merge branch '4.4' into fix-tx-timeout-rounding-4.4

robsdedude committed:
Adjust unit tests

For better backwards compatibility, we're sending negative timeouts to the
DBMS as the 4.4 driver did before (even though those values might not have the
intuitive effect, we don't want to change the behavior of the driver too much in
a patch release).
